src/Uno.UI/UI/Xaml/UIElement.debugging.cs

```diff
@@ -0,0 +1,205 @@
+		/// <remarks>
+		/// Currently won't work with value-typed resources (eg double, Thickness) since it uses ReferencEquals() and they will be boxed.
+		/// </remarks>
+		internal object[] GetKeysForResource(object resource)
+		{
+			return Inner().ToArray();
+
+			IEnumerable<object> Inner()
+			{
+				var fe = this as FrameworkElement;
+				while (fe != null)
+				{
+					foreach (var key in TryFindResource(fe.Resources))
+					{
+						yield return key;
+					}
+
+					fe = fe.Parent as FrameworkElement;
+				}
+
+				foreach (var key in TryFindResource(Application.Current.Resources))
+				{
+					yield return key;
+				}
+				foreach (var key in TryFindResource(Uno.UI.GlobalStaticResources.MasterDictionary))
+				{
+					yield return key;
+				}
+
+				IEnumerable<object> TryFindResource(ResourceDictionary resourceDictionary)
+				{
+					foreach (var kvp in resourceDictionary)
+					{
+						if (ReferenceEquals(resource, kvp.Value)) // TODO: doesn't work for value types
+						{
+							yield return kvp.Key;
+						}
+					}
+
+					foreach (var mergedDict in resourceDictionary.MergedDictionaries)
+					{
+						foreach (var key in TryFindResource(mergedDict))
+						{
+							yield return key;
+						}
+					}
+
+					foreach (var themeDict in resourceDictionary.ThemeDictionaries.Values.OfType<ResourceDictionary>())
+					{
+						foreach (var key in TryFindResource(themeDict))
+						{
+							yield return key;
+						}
+					}
+				}
+			}
+		}
+
+		/// <summary>
+		/// Debugging helper method to get the parent element that this element is inheriting <paramref name="propertyName"/> from.
+		///
+		/// Eg, if you call textBlock.GetPropInheritanceParent("Foreground"), it will return the first parent which has a property called
+		/// "Foreground" (most likely the Control whose template contains the TextBlock).
+		/// </summary>
+		internal FrameworkElement GetPropInheritanceParent(string propertyName)
+		{
+			var dependencyProp = GetDPByName(propertyName);
+			if (!IsInherited(dependencyProp, GetType()))
+			{
+				return null;
+			}
+
+			return Uno.UI.Extensions.DependencyObjectExtensions.FindFirstParent<FrameworkElement>(this, fe => IsInherited(GetDPByName(propertyName, fe), fe.GetType()), includeCurrent: false);
+
+			bool IsInherited(DependencyProperty dp, Type type)
+			{
+				if (dp.Name == "DataContext")
+				{
+					return true;
+				}
+				var metadata = dp?.GetMetadata(type) as FrameworkPropertyMetadata;
+				return metadata?.Options.HasFlag(FrameworkPropertyMetadataOptions.Inherits) ?? false;
+			}
+		}
+
+		/// <summary>
+		/// Debugging helper method to get a condensed summary of visual states on controls in this element's hierarchy.
+		/// </summary>
+		internal string GetVisualStatesSummary()
+		{
+			var sb = new StringBuilder();
+			foreach (var ancestor in Uno.UI.Extensions.DependencyObjectExtensions.GetAllParents(this))
+			{
+				if (ancestor is Control control && control.GetTemplateRoot() is { } root)
```

Comment:
```suggestion
				if (ancestor is Control control && control.GetTemplateRoot() is IFrameworkElement root)
```

Reply:
it's not necessary - `GetTemplateRoot()` is already returning `IFrameworkElement`